File: src/main/java/com/example/day04/mianshi_02_07/Solution.java
package com.example.day04.mianshi_02_07;

import com.example.day04.leetcode_24.ListNode;

public class Solution {
    public ListNode getIntersectionNode(ListNode headA,ListNode headB){
        ListNode curA = headA;
        ListNode curB = headB;
        int lenA = 0,lenB = 0;
        while(curA != null){
            lenA ++;
            curA = curA.next;
        }

        while(curB != null){
            lenB ++;
            curB = curB.next;
        }

        curA = headA;
        curB = headB;
        if(lenB > lenA){
            int tmp = lenA;
            lenA = lenB;
            lenB = tmp;
            ListNode tmpNode = curA;
            curA = curB;
            curB = tmpNode;
        }

        int gap = lenA - lenB;
        while(gap > 0){
            curA = curA.next;
            gap --;
        }

        while(curA != null){
            if(curA == curB){
                return curA;
            }
            curA = curA.next;
            curB = curB.next;
        }
        return null;
    }
}

File: src/main/java/com/example/day04/mianshi_02_07/Solution.py
from typing import Optional


class ListNode:
    def __init__(self,x):
        self.val = x
        self.next = None
class Solution:
    def getIntersectionNode(self,headA: Optional[ListNode],headB: Optional[ListNode]):
        if not headA or not headB:
            return None
        pa = headA
        pb = headB
        lenA = lenB = 0
        while pa:
            lenA += 1
            pa = pa.next
        while pb:
            lenB += 1
            pb = pb.next
        if lenB > lenA:
            tmp = headA
            headA = headB
            headB = tmp
            cur =lenA
            lenA = lenB
            lenB = cur
        pa = headA
        pb = headB
        for i in range(lenA - lenB):
            pa = pa.next
        while pa and pb:
            if pa == pb:
                return pa
            pa = pa.next
            pb = pb.next
        return None